File: src/main/java/com/github/nsc/de/shake/util/Characters.java
package com.github.nsc.de.shake.util;


public class Characters {

    public static boolean isHexCharacter(char c) {

        //       48-57 (0-9)         65-70 (A-F)        97-102 (a-f)
        return c > 47 && c < 58 || c > 64 && c < 71 || c > 96 && c < 103;

    }

    public static boolean isNumberCharacter(char c) {
        //       48-57 (0-9)
        return c > 47 && c < 58;
    }

    public static boolean isNumberOrDotCharacter(char c) {
        //       48-57 (0-9)       46 ('.')
        return c > 47 && c < 58 || c == 46;
    }

    public static boolean isWhitespaceCharacter(char c) {
        //      (' ')     ('\t')    ('\r')
        return c == 32 || c == 9 || c == 13;
    }

    public static boolean isIdentifierCharacter(char c) {

        //       48-57 (0-9)         65-90 (A-Z)         97-122 (a-z)        '_'
        return c > 47 && c < 58 || c > 64 && c < 91 || c > 96 && c < 123 || c == 95;

    }

    public static boolean isIdentifierStartCharacter(char c) {

        //       65-90 (A-Z)         97-122 (a-z)        '_'
        return c > 64 && c < 91 || c > 96 && c < 123 || c == 95;

    }


    public static String parseString(String s) {

        StringBuilder string = new StringBuilder();
        int length = s.length();
        for(int i = 0; i < length; i++) {
            char c = s.charAt(i);
            if(c == '\\') {
                switch(s.charAt(++i)) {
                    case 't': string.append('\t'); break;
                    case 'b': string.append('\b'); break;
                    case 'n': string.append('\n'); break;
                    case 'r': string.append('\r'); break;
                    case 'f': string.append('\f'); break;
                    case '\'': string.append('\''); break;
                    case '"': string.append('\"'); break;
                    case '\\': string.append('\\'); break;
                    case 'u':
                        StringBuilder unicode = new StringBuilder();
                        int to = i + 5;
                        for(; i < to; ++i) {
                            c = s.charAt(i);
                            if(!isHexCharacter(c)) throw new Error("Expecting hex char");
                            unicode.append(c);
                        }
                        string.append(Integer.parseInt(unicode.toString(), 16));
                        break;
                    default:
                        throw new Error();

                }
            }
            else string.append(c);

        }
        return string.toString();
    }

}

File: util/src/commonMain/kotlin/com/github/shakelang/shake/util/Characters.kt
package com.github.shakelang.shake.util

import com.github.shakelang.shake.util.json.json
import kotlin.js.JsName
import kotlin.jvm.JvmStatic


/**
 * Utilities for wirking with characters
 *
 * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
 */
object Characters {

    /**
     * Is the given character a hex character (0-f)
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    @JvmStatic
    @JsName("isHexCharacter")
    fun isHexCharacter(c: Char): Boolean = c in '0'..'9' || c in 'A'..'F' || c in 'a'..'f'

    /**
     * Is the given character a number character (0-9)
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    @JvmStatic
    @JsName("isNumberCharacter")
    fun isNumberCharacter(c: Char): Boolean = c in '0'..'9'

    /**
     * Is the given character a number character (0-9) or a dot character
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    @JvmStatic
    @JsName("isNumberOrDotCharacter")
    fun isNumberOrDotCharacter(c: Char): Boolean = c in '0'..'9' || c == '.'

    /**
     * Is the given character a ignored whitespace character ('\r', '\t' or ' ', **not '\n'**)
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    @JvmStatic
    @JsName("isWhitespaceCharacter")
    fun isWhitespaceCharacter(c: Char): Boolean = c == '\r' || c == '\t' || c == ' '

    /**
     * Is the given character a part of an identifier (A-Z, a-z, 0-9 and '_')
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    @JvmStatic
    @JsName("isIdentifierCharacter")
    fun isIdentifierCharacter(c: Char): Boolean = c in '0'..'9' || c in 'A'..'Z' || c in 'a'..'z' || c == '_'

    /**
     * Is the given character the start of a identifier (A-Z, a-z and '_', **not 0-9**)
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    @JvmStatic
    @JsName("isIdentifierStartCharacter")
    fun isIdentifierStartCharacter(c: Char): Boolean = c in 'A'..'Z' || c in 'a'..'z' || c == '_'


    /**
     * Parses string contents. Not including string start & end
     * e.g.
     *
     * `"hello world"` → `"hello world"`
     *
     * `"hello\tworld"` → `"hello world"`
     *
     * `"\u0021"` → `"!"`
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    @JvmStatic
    @JsName("parseString")
    fun parseString(s: String): String {

        val string = StringBuilder()

        // We use a while loop because kotlin has no normal for loop
        var i = 0
        while (i < s.length) {
            var c = s[i]

            // Escape sequences
            if (c == '\\') {
                when (s[++i]) {
                    't' -> string.append('\t')
                    'b' -> string.append('\b')
                    'n' -> string.append('\n')
                    'r' -> string.append('\r')
                    '\'' -> string.append('\'')
                    '"' -> string.append('\"')
                    '\\' -> string.append('\\')
                    'u' -> {

                        // parse unicode

                        // create string for storing the unicode
                        val unicode = StringBuilder()

                        // Get the next 4 hex characters
                        val to = i + 4
                        while (i < to) {
                            c = s[++i]

                            // Throw an error if the character is not a hex character
                            if (!isHexCharacter(c)) throw Error("Expecting hex char, got $c")

                            unicode.append(c)

                        }

                        // Parse the unicode and append it to the string
                        string.append(unicode.toString().toInt(16).toChar())
                    }

                    // When we don't know the escape sequence we throw an error
                    else -> throw Error("Unknown escape sequence '\\${s[i]}'")
                }


            }

            // If the next character is not a escape sequence we
            // just append it to the string result
            else string.append(c)

            // Increase the counting variable
            i++
        }

        // Return the generated string
        return string.toString()
    }
}